### ISSUE TITLE
Parsing Error "Await only valid in Async Function" with all User JS Scripts

### ISSUE THREAD

**Issue by morgen1990** [SilentVoid13/Templater#1016]
**Plugin information (please complete the following information):**

-   OS: Mac Ventura 13.1
-   Templater version: 1.16.0
-   Obsidian version: 1.1.9
-   Templater settings: Template folder location: "Templates/", Timeout: "5" (add the templates not working with their associated command)]

My templates are stored in the /Templates folder, javascript files are stored in the /Templates/Javascript Files folder and the Templater plugin settings are updated to reflect this.

**Describe the bug**
When I try to run user javascript files as described on the [how to use templater js scripts site](https://shbgm.ca/blog/obsidian/how-to-use-templater-js-scripts), I get the following error message:

" Templater Error: Template parsing error, aborting.
await is only valid in async functions and the top level bodies of modules"

The truly bizarre thing is that I get this error even when trying to run scripts that don't contain await or any async functions. I've tried lots of different scripts to troubleshoot the issue, copied three examples below. 

# Example 1: test2.js

The JS File: 
```javascript
async function test2(tp) {
    const text = await tp.system.prompt("What's Good?")
    new Notice(text, 5000)
}

module.exports = test2
```
![Screenshot 2023-01-31 at 9 06 35 AM](https://user-images.githubusercontent.com/113865048/215679821-14613022-ff77-4eb9-8c11-f29eb6209d34.png)

The Templater Command File: test2
<%* tp.user.test2(tp) %>
<img width="270" alt="Screenshot 2023-01-31 at 9 07 15 AM" src="https://user-images.githubusercontent.com/113865048/215679934-408ae2e0-b981-4540-909b-c407b92824c1.png">

The Error Message

![Screenshot 2023-01-31 at 8 42 09 AM](https://user-images.githubusercontent.com/113865048/215679563-dd4975e1-0213-48c9-aeec-34f4f0c7b832.png)

# Example 2

The JS File: EditFolderMetadata.js
```javascript

const {update} = app.plugins.plugins["metaedit"].api
const root = app.vault.getRoot()
const folders = root.children.filter(child => child.children)
const selectedFolder = await tp.system.suggester(e => e.name, folders, false, "Choose a Folder")
const metaKey = await tp.system.prompt("What Meta Key?")
const metaValue = await tp.system.prompt("Meta Value")
if (selectedFolder.children) {
selectedFolder.children.forEach(async (child) => {
    const {frontmatter} = app.metadataCache.getCache(child.path)
	const content = await app.vault.read(child)
	if (frontmatter) {
		if (Object.keys(frontmatter).includes(metaKey)) {
			update(metaKey, metaValue, child)
		} else {
			const contentArray = content.split("\n")
			contentArray.splice(1, 0, `${metaKey}: ${metaValue}`)
			await app.vault.modify(child, contentArray.join("\n"))
		}
	} else {
		const updatedContent = `---\n${metaKey}: ${metaValue}\n---`.concat(content)
		await app.vault.modify(child, updatedContent)
	}
})
} else {
new Notice("No Notes in Selected Folder")
}

module.exports = EditFolderMetadata.js
```
![Screenshot 2023-01-31 at 9 09 05 AM](https://user-images.githubusercontent.com/113865048/215680244-03a7683f-bad5-4d3b-ab76-70a8f8e79e95.png)

The Templater Command File: EditFolderMetadata
<%* tp.user.EditFolderMetadata(tp) %>

<img width="362" alt="Screenshot 2023-01-31 at 9 09 58 AM" src="https://user-images.githubusercontent.com/113865048/215680378-70469602-68ff-43c2-9f86-d574c5c0bf3e.png">

The Error Message
![Screenshot 2023-01-31 at 8 42 09 AM](https://user-images.githubusercontent.com/113865048/215679563-dd4975e1-0213-48c9-aeec-34f4f0c7b832.png)

# Example 3: myscript.js

The JS File: 
```javascript
function my_function (msg) {
    return `Message from my script: ${msg}`;
}
module.exports = my_function;

```

![Screenshot 2023-01-31 at 9 12 51 AM](https://user-images.githubusercontent.com/113865048/215680905-0fbf72ed-8973-41dc-b8ac-dd4a8a1452db.png)

The Templater Command File: myscript
<% tp.user.myscript(msg) %>

<img width="301" alt="Screenshot 2023-01-31 at 9 13 37 AM" src="https://user-images.githubusercontent.com/113865048/215681049-93c2d994-f0e9-4b74-8ab4-879e13d15383.png">

The Error Message

<img width="653" alt="Screenshot 2023-01-31 at 9 14 03 AM" src="https://user-images.githubusercontent.com/113865048/215681108-a1036989-b467-42d9-b2b3-5e64ba91f926.png">

Even with example 3, which doesn't have any async functions I keep getting the same error. 

Any suggestions or advice would be much appreciated.

**Comment by welpdx** [comment]
# 1. test2: Works fine 


<details>
           <summary style="font-size:20">test2 Demo</summary>
           <img src="https://i.imgur.com/UZMycF7.gif" alt="demo" />
</details>


no errors in console or anything...
I have my scripts folder in templates folder too.

# 2. metaFolder: Works fine after fixes
works fine after making 2 changes:
1) change last line to 
```js
module.exports = EditFolderMetadata
```

2) enclose everything into a function:
```js
async function EditFolderMetadata(tp){
...
}
```


<details><summary style="font-size:20">Final code looks like this:</summary>
<p>

```js
async function EditFolderMetadata(tp){
	const {update} = app.plugins.plugins["metaedit"].api
	const root = app.vault.getRoot()
	const folders = root.children.filter(child => child.children)
	const selectedFolder = await tp.system.suggester(e => e.name, folders, false, "Choose a Folder")
	const metaKey = await tp.system.prompt("What Meta Key?")
	const metaValue = await tp.system.prompt("Meta Value")
	if (selectedFolder.children) {
	selectedFolder.children.forEach(async (child) => {
		const {frontmatter} = app.metadataCache.getCache(child.path)
		const content = await app.vault.read(child)
		if (frontmatter) {
			if (Object.keys(frontmatter).includes(metaKey)) {
				update(metaKey, metaValue, child)
			} else {
				const contentArray = content.split("\n")
				contentArray.splice(1, 0, `${metaKey}: ${metaValue}`)
				await app.vault.modify(child, contentArray.join("\n"))
			}
		} else {
			const updatedContent = `---\n${metaKey}: ${metaValue}\n---`.concat(content)
			await app.vault.modify(child, updatedContent)
		}
	})
	} else {
	new Notice("No Notes in Selected Folder")
	}
}

module.exports = EditFolderMetadata
```

</p>
</details>


<details>
           <summary style="font-size:20">metaFolder Demo</summary>
           <img src="https://i.imgur.com/AKBn4sB.gif" alt="demo" />
</details>


The errors in console is unrelated to the user script. It is because I think I dont have folder notes set up or something.


# 3. myscript: Works fine
this one works after 1 fix:
1) when calling the user function, msg is not defined. I just changed that into a string so that it accepts. 
```js
//instead of
<% tp.user.myscript(msg) %>
//I did
<% tp.user.myscript("msg") %>
```


<details>
           <summary style="font-size:20">myscript Demo</summary>
           <img src="https://i.imgur.com/7nUGpYv.gif" alt="demo" />
</details>




## tl;dr
Everything looks good. Idk why the first one didn't work for you. 
To do these tests, I have created a new vault and installed only Templater, and MetaEdit.
Other than that, I could not replicate the errors you have encountered. 

I've noticed that you have the same errors each time and I think it might help if you press `F5` in console to restart obsidian from time to time. This refreshes the caches in obsidian so that your changes to your templates and code will be made each time you execute your templater script. 

~welp

**Comment by kenfeliciano** [comment]
I was having this same issue which brought me here. In example 3, with no async functions, getting the same error lead me to figuring out I had ***other*** scripts in my scripts folder which Templater was trying to read and process. They were not compatible. I created a new folder for only Templater scripts. Now the same script worked just fine.

In Templater settings, it does say, "All JavaScript files in this folder will be loaded as CommonJS modules, to import custom user functions." I didn't read that very closely or understand the ramifications of that.
